# Intro to Pandas
#### Illinois Data Science Club

This notebook will cover the basic facets of Python necessary for data science, as well as specific features of the pandas library. Given our time constraints, we won't be covering everything there is to know about Python, but we are confident that plenty is covered in these workshops for a successful data dive project to blossom. If you'd like to go above and beyond, documentation for the latest versions of [Python](https://docs.python.org/3/) and [pandas](https://pandas.pydata.org/docs/) are linked.

## Python

Python is a programming language featuring constructs like variables and functions. Variables, much like in math, are objects that hold some value. For example,

In [ ]:
# make x

...

Once we've made a variable, we can call it later in the code.

In [34]:
# print x

print(...)

Ellipsis


We just used `print`, a function! Functions are how we change variables and turn input into output. Functions typically take arguments in the form of variables. For example, the `print` function takes one variable as input and prints it out to the user.

There are quite a few different data types, all of which you will frequently see in your data analysis.

In [4]:
var1 = 'I love iDSC!'
type(var1)

str

We can use `type` to check the data type of a variable.

In [5]:
var2 = 67
type(var2)

int

In [6]:
var3 = 3.14
type(var3)

float

In [44]:
# make a boolean

idsc_is_the_best_data_science_organization_on_campus = ...
type(idsc_is_the_best_data_science_organization_on_campus)

ellipsis

We can compare variables using operations like `==`, `>`, `>=`, etc.

In [8]:
x = 10
y = 20

In [36]:
x == 10

True

In [38]:
x == 50

False

We can compare two variables as well.

In [ ]:
# compare x and y

...

False

Understanding of these operators comes with lots of time and practice. A nice cheat sheet can be found [here](https://learnpython.com/blog/python-operators-cheat-sheet/).

## Pandas

Now to get into some actual data science! Data science in Python is made much easier with the pandas library, a package of functions and features that allow us to import, format and transform data into an ideal format for modeling and visualizing.

We start by importing pandas (if for some reason you don't have pandas installed, you will get an error. Try running `pip install pandas`, `pip3 install pandas`, or `python3 install pandas` in your terminal). We typically import pandas "as pd", which means that anywhere we use a pandas feature, we can just write "pd" instead of "pandas". 

In [ ]:
# import pandas!!!!

...

We use the `read_csv` function to turn a `.csv` file into a pandas dataframe. This particular dataset contains the current batting leaders in the Women's Professional Baseball League 2026 season and was acquired [here](https://www.baseball-reference.com/register/leader.cgi?type=bat&id=b910b711#all_league_batting).

In [ ]:
# reads a .csv file as a dataframe

df = ...

Now that we have our data in a dataframe, we can use pandas to explore it a little bit.

In [ ]:
# displays the first 5 rows of the dataframe

...

,Rk,Name,Age,Tm,Lev,Aff,G,PA,AB,R,...,OBP,SLG,OPS,TB,GDP,HBP,SH,SF,IBB,Name-additional
0,1,Denae Benites,24,NYH,Wmn,NaN,13,56,48,20,...,0.589,1.125,1.714,54,NaN,1,0,1,2,player.fcgi?id=benite000den
1,2,Suzuka Yamamoto*,27,BOS,Wmn,NaN,7,19,15,1,...,0.632,0.667,1.298,10,NaN,1,0,0,0,player.fcgi?id=yamamo000suz
2,3,Kelsie Whitmore#,27,SFF,Wmn,NaN,13,55,44,20,...,0.564,1.205,1.768,53,NaN,1,0,1,2,player.fcgi?id=whitmo000kel
3,4,Ashton Lansdell,25,LAQ,Wmn,NaN,13,52,38,18,...,0.615,0.868,1.484,33,NaN,3,0,0,1,player.fcgi?id=lansde000ash
4,5,Elodie Ciamarro,19,NYH,Wmn,NaN,8,24,24,5,...,0.458,0.542,1.000,13,NaN,0,0,0,0,player.fcgi?id=ciamar000elo


In [14]:
# displays the last 5 rows of the dataframe

df.tail()

,Rk,Name,Age,Tm,Lev,Aff,G,PA,AB,R,...,OBP,SLG,OPS,TB,GDP,HBP,SH,SF,IBB,Name-additional
44,45,Raine Padgham,20,BOS,Wmn,NaN,12,35,30,5,...,0.314,0.200,0.514,6,NaN,0,0,0,0,player.fcgi?id=padgha000rai
45,46,Ela Day-Bédard,21,SFF,Wmn,NaN,10,29,23,6,...,0.310,0.174,0.484,4,NaN,4,0,1,0,player.fcgi?id=day-be000ela
46,47,Hyeonah Kim,26,BOS,Wmn,NaN,10,26,18,8,...,0.400,0.222,0.622,4,NaN,4,1,0,0,player.fcgi?id=kim---021hye
47,48,Diana Ibarra,26,NYHLAQ,Wmn,NaN,11,27,22,1,...,0.296,0.136,0.433,3,NaN,2,0,0,0,player.fcgi?id=ibarra000dia
48,49,Jaida Lee,20,NYH,Wmn,NaN,10,15,13,3,...,0.200,0.077,0.277,1,NaN,0,0,0,0,player.fcgi?id=lee---000jai


In [ ]:
# displays the number of rows and columns

...

(49, 30)

This means there are 49 rows and 30 columns (or, 49 observations and 30 variables).

In [ ]:
# provides summary statistics for each column

...

,Rk,Age,Aff,G,PA,AB,R,H,2B,3B,...,BA,OBP,SLG,OPS,TB,GDP,HBP,SH,SF,IBB
count,49.00000,49.000000,0.0,49.000000,49.000000,49.000000,49.000000,49.000000,49.000000,49.0,...,49.000000,49.000000,49.000000,49.000000,49.000000,0.0,49.000000,49.000000,49.000000,49.000000
mean,25.00000,23.387755,NaN,10.795918,35.979592,29.632653,7.673469,10.061224,1.734694,0.0,...,0.328857,0.433286,0.465980,0.899204,14.673469,NaN,1.326531,0.122449,0.408163,0.102041
std,14.28869,3.812350,NaN,1.989342,12.157799,10.240227,4.913365,5.421440,1.551552,0.0,...,0.102563,0.094044,0.227057,0.307790,11.051749,NaN,1.505376,0.389051,0.642857,0.420560
min,1.00000,18.000000,NaN,6.000000,15.000000,13.000000,1.000000,1.000000,0.000000,0.0,...,0.077000,0.200000,0.077000,0.277000,1.000000,NaN,0.000000,0.000000,0.000000,0.000000
25%,13.00000,20.000000,NaN,10.000000,26.000000,22.000000,4.000000,6.000000,1.000000,0.0,...,0.270000,0.360000,0.318000,0.679000,7.000000,NaN,0.000000,0.000000,0.000000,0.000000
50%,25.00000,23.000000,NaN,11.000000,36.000000,30.000000,7.000000,9.000000,1.000000,0.0,...,0.333000,0.421000,0.421000,0.851000,12.000000,NaN,1.000000,0.000000,0.000000,0.000000
75%,37.00000,25.000000,NaN,12.000000,45.000000,38.000000,10.000000,13.000000,3.000000,0.0,...,0.406000,0.500000,0.592000,1.081000,17.000000,NaN,2.000000,0.000000,1.000000,0.000000
max,49.00000,34.000000,NaN,13.000000,58.000000,49.000000,20.000000,26.000000,6.000000,0.0,...,0.542000,0.632000,1.205000,1.768000,54.000000,NaN,6.000000,2.000000,3.000000,2.000000


What if we want to examine a specific column?

In [ ]:
# returns only the "Tm" column of df

...

0        NYH
1        BOS
2        SFF
3        LAQ
4        NYH
5        SFF
6        NYH
7        SFF
8        SFF
9        LAQ
10       SFF
11       NYH
12       SFF
13       NYH
14       LAQ
15       BOS
16       LAQ
17       SFF
18       LAQ
19       LAQ
20       SFF
21       SFF
22       LAQ
23       LAQ
24       BOS
25       NYH
26       SFF
27       BOS
28       LAQ
29       NYH
30       NYH
31       BOS
32       NYH
33       NYH
34       LAQ
35       BOS
36       BOS
37       LAQ
38       NYH
39       BOS
40       BOS
41       LAQ
42       BOS
43       BOS
44       BOS
45       SFF
46       BOS
47    NYHLAQ
48       NYH
Name: Tm, dtype: str

In [ ]:
# displays how many players each team has

...

Tm
BOS       13
NYH       12
LAQ       12
SFF       11
NYHLAQ     1
Name: count, dtype: int64

I will now introduce one of the most useful features in pandas; filtering, which starts out simple but can get quite complex. For example, what if we want to find all players with batting averages above .400?

In [ ]:
# gets all players with BA > 0.350

df[...]

,Rk,Name,Age,Tm,Lev,Aff,G,PA,AB,R,...,OBP,SLG,OPS,TB,GDP,HBP,SH,SF,IBB,Name-additional
0,1,Denae Benites,24,NYH,Wmn,NaN,13,56,48,20,...,0.589,1.125,1.714,54,NaN,1,0,1,2,player.fcgi?id=benite000den
1,2,Suzuka Yamamoto*,27,BOS,Wmn,NaN,7,19,15,1,...,0.632,0.667,1.298,10,NaN,1,0,0,0,player.fcgi?id=yamamo000suz
2,3,Kelsie Whitmore#,27,SFF,Wmn,NaN,13,55,44,20,...,0.564,1.205,1.768,53,NaN,1,0,1,2,player.fcgi?id=whitmo000kel
3,4,Ashton Lansdell,25,LAQ,Wmn,NaN,13,52,38,18,...,0.615,0.868,1.484,33,NaN,3,0,0,1,player.fcgi?id=lansde000ash
4,5,Elodie Ciamarro,19,NYH,Wmn,NaN,8,24,24,5,...,0.458,0.542,1.000,13,NaN,0,0,0,0,player.fcgi?id=ciamar000elo
5,6,Skylar Kaplan*,24,SFF,Wmn,NaN,12,50,44,14,...,0.520,0.614,1.134,27,NaN,0,0,0,0,player.fcgi?id=kaplan000sky
6,7,Claire O'Sullivan,31,NYH,Wmn,NaN,11,26,20,5,...,0.500,0.600,1.100,12,NaN,1,0,2,0,player.fcgi?id=osulli000cla
7,8,Alexia Jorge,22,SFF,Wmn,NaN,11,45,34,14,...,0.578,0.647,1.225,22,NaN,6,0,0,0,player.fcgi?id=jorge-001ale
8,9,Samantha Gutierrez,23,SFF,Wmn,NaN,6,21,16,6,...,0.571,0.688,1.259,11,NaN,2,0,0,0,player.fcgi?id=gutier002sam
9,10,Samaria Benítez,23,LAQ,Wmn,NaN,12,42,38,8,...,0.476,0.421,0.897,16,NaN,1,0,0,0,player.fcgi?id=benite000sam


In [ ]:
# gets all players on the team NYH

df[...]

,Rk,Name,Age,Tm,Lev,Aff,G,PA,AB,R,...,OBP,SLG,OPS,TB,GDP,HBP,SH,SF,IBB,Name-additional
0,1,Denae Benites,24,NYH,Wmn,NaN,13,56,48,20,...,0.589,1.125,1.714,54,NaN,1,0,1,2,player.fcgi?id=benite000den
4,5,Elodie Ciamarro,19,NYH,Wmn,NaN,8,24,24,5,...,0.458,0.542,1.000,13,NaN,0,0,0,0,player.fcgi?id=ciamar000elo
6,7,Claire O'Sullivan,31,NYH,Wmn,NaN,11,26,20,5,...,0.500,0.600,1.100,12,NaN,1,0,2,0,player.fcgi?id=osulli000cla
11,12,Natsuki Yonetani,24,NYH,Wmn,NaN,13,58,49,19,...,0.500,0.592,1.092,29,NaN,2,0,0,0,player.fcgi?id=yoneta000nat
13,14,London Studer*,19,NYH,Wmn,NaN,12,39,36,7,...,0.410,0.389,0.799,14,NaN,0,0,1,0,player.fcgi?id=studer000lon
25,26,Kylee Lahners,33,NYH,Wmn,NaN,12,44,28,7,...,0.545,0.536,1.081,15,NaN,6,0,1,0,player.fcgi?id=lahner000kyl
29,30,Alyssa Zettlemoyer,19,NYH,Wmn,NaN,12,33,26,3,...,0.424,0.385,0.809,10,NaN,4,0,1,0,player.fcgi?id=zettle000aly
30,31,Claire Eccles*,28,NYH,Wmn,NaN,11,34,30,7,...,0.313,0.367,0.679,11,NaN,0,2,1,0,player.fcgi?id=eccles000cla
32,33,Keira Izumi,19,NYH,Wmn,NaN,11,29,24,4,...,0.379,0.458,0.838,11,NaN,2,0,1,0,player.fcgi?id=izumi-001kei
33,34,Madison Willan,25,NYH,Wmn,NaN,10,24,18,3,...,0.417,0.389,0.806,7,NaN,0,0,1,0,player.fcgi?id=willan000mad


We can also save these dataframes in other variables for later use.

In [ ]:
# gets all players under 20 with at least 1 HR

df_filtered = df[...]      # note the parentheses!!

In [22]:
df_filtered

,Rk,Name,Age,Tm,Lev,Aff,G,PA,AB,R,...,OBP,SLG,OPS,TB,GDP,HBP,SH,SF,IBB,Name-additional
32,33,Keira Izumi,19,NYH,Wmn,NaN,11,29,24,4,...,0.379,0.458,0.838,11,NaN,2,0,1,0,player.fcgi?id=izumi-001kei
37,38,Thaima Maximiliana#,19,LAQ,Wmn,NaN,12,28,25,4,...,0.321,0.400,0.721,10,NaN,1,0,0,0,player.fcgi?id=maximi000tha


Finally, there are some interesting (self-explanatory) functions to find summary statistics of specific columns.

In [23]:
df['Age'].max()

np.int64(34)

In [24]:
df['Age'].min()

np.int64(18)

In [25]:
df['Age'].mean()

np.float64(23.387755102040817)

In [26]:
df['Age'].median()

np.float64(23.0)

For getting minimums and maximums, the following are usually preferred.

In [ ]:
# get the 5 oldest players

df.nlargest(...)

,Rk,Name,Age,Tm,Lev,Aff,G,PA,AB,R,...,OBP,SLG,OPS,TB,GDP,HBP,SH,SF,IBB,Name-additional
38,39,Val Perez,34,NYH,Wmn,NaN,7,29,25,9,...,0.310,0.280,0.590,7,NaN,1,0,1,0,player.fcgi?id=perez-004val
25,26,Kylee Lahners,33,NYH,Wmn,NaN,12,44,28,7,...,0.545,0.536,1.081,15,NaN,6,0,1,0,player.fcgi?id=lahner000kyl
6,7,Claire O'Sullivan,31,NYH,Wmn,NaN,11,26,20,5,...,0.500,0.600,1.100,12,NaN,1,0,2,0,player.fcgi?id=osulli000cla
22,23,Sarah Edwards#,30,LAQ,Wmn,NaN,11,41,33,6,...,0.463,0.455,0.918,15,NaN,1,0,0,0,player.fcgi?id=edward000sar
26,27,Amanda Gianelloni,29,SFF,Wmn,NaN,13,56,41,15,...,0.436,0.415,0.851,17,NaN,2,1,3,0,player.fcgi?id=gianel000ama


In [ ]:
# get the 5 youngest players

df.nsmallest(...)

,Rk,Name,Age,Tm,Lev,Aff,G,PA,AB,R,...,OBP,SLG,OPS,TB,GDP,HBP,SH,SF,IBB,Name-additional
23,24,Adelaide Frank,18,LAQ,Wmn,NaN,7,17,15,2,...,0.412,0.333,0.745,5,NaN,0,0,0,0,player.fcgi?id=frank-000ade
41,42,Amira Hondras*,18,LAQ,Wmn,NaN,10,38,31,4,...,0.342,0.290,0.632,9,NaN,1,0,1,0,player.fcgi?id=hondra000ami
42,43,Maïka Dumais*,18,BOS,Wmn,NaN,12,28,23,7,...,0.357,0.261,0.618,6,NaN,0,0,0,0,player.fcgi?id=dumais000mai
4,5,Elodie Ciamarro,19,NYH,Wmn,NaN,8,24,24,5,...,0.458,0.542,1.000,13,NaN,0,0,0,0,player.fcgi?id=ciamar000elo
13,14,London Studer*,19,NYH,Wmn,NaN,12,39,36,7,...,0.410,0.389,0.799,14,NaN,0,0,1,0,player.fcgi?id=studer000lon


## Practice on your own!

In [ ]:
# What is the median AB in the league? (column AB represents at-bats)

...

np.float64(30.0)

In [ ]:
# What are the 7 players with the highest SLG? (column SLG represents slugging)

...

,Rk,Name,Age,Tm,Lev,Aff,G,PA,AB,R,...,OBP,SLG,OPS,TB,GDP,HBP,SH,SF,IBB,Name-additional
2,3,Kelsie Whitmore#,27,SFF,Wmn,NaN,13,55,44,20,...,0.564,1.205,1.768,53,NaN,1,0,1,2,player.fcgi?id=whitmo000kel
0,1,Denae Benites,24,NYH,Wmn,NaN,13,56,48,20,...,0.589,1.125,1.714,54,NaN,1,0,1,2,player.fcgi?id=benite000den
3,4,Ashton Lansdell,25,LAQ,Wmn,NaN,13,52,38,18,...,0.615,0.868,1.484,33,NaN,3,0,0,1,player.fcgi?id=lansde000ash
16,17,Jamie Mackay,23,LAQ,Wmn,NaN,13,46,40,14,...,0.457,0.800,1.257,32,NaN,4,0,0,0,player.fcgi?id=mackay000jam
24,25,Lexi Hastings*,23,BOS,Wmn,NaN,10,32,24,7,...,0.500,0.708,1.208,17,NaN,1,0,0,0,player.fcgi?id=hastin000lex
8,9,Samantha Gutierrez,23,SFF,Wmn,NaN,6,21,16,6,...,0.571,0.688,1.259,11,NaN,2,0,0,0,player.fcgi?id=gutier002sam
12,13,Jill Albayati,22,SFF,Wmn,NaN,9,39,32,11,...,0.513,0.688,1.200,22,NaN,1,0,0,0,player.fcgi?id=albaya000jil


In [ ]:
# Find all players on team LAQ who have played at least 10 games (column G represents games played)

...

,Rk,Name,Age,Tm,Lev,Aff,G,PA,AB,R,...,OBP,SLG,OPS,TB,GDP,HBP,SH,SF,IBB,Name-additional
3,4,Ashton Lansdell,25,LAQ,Wmn,NaN,13,52,38,18,...,0.615,0.868,1.484,33,NaN,3,0,0,1,player.fcgi?id=lansde000ash
9,10,Samaria Benítez,23,LAQ,Wmn,NaN,12,42,38,8,...,0.476,0.421,0.897,16,NaN,1,0,0,0,player.fcgi?id=benite000sam
14,15,Caitlin Eynon*,23,LAQ,Wmn,NaN,10,34,31,7,...,0.412,0.613,1.025,19,NaN,0,0,1,0,player.fcgi?id=eynon-000cai
16,17,Jamie Mackay,23,LAQ,Wmn,NaN,13,46,40,14,...,0.457,0.800,1.257,32,NaN,4,0,0,0,player.fcgi?id=mackay000jam
19,20,Michelle Roche,20,LAQ,Wmn,NaN,10,19,17,4,...,0.421,0.529,0.950,9,NaN,0,0,0,0,player.fcgi?id=roche-003mic
22,23,Sarah Edwards#,30,LAQ,Wmn,NaN,11,41,33,6,...,0.463,0.455,0.918,15,NaN,1,0,0,0,player.fcgi?id=edward000sar
28,29,Maggie Foxx,20,LAQ,Wmn,NaN,13,46,39,11,...,0.413,0.385,0.798,15,NaN,0,0,0,0,player.fcgi?id=foxx--000mag
34,35,Mo'ne Davis,25,LAQ,Wmn,NaN,11,37,33,7,...,0.351,0.364,0.715,12,NaN,0,0,0,0,player.fcgi?id=davis-002mon
37,38,Thaima Maximiliana#,19,LAQ,Wmn,NaN,12,28,25,4,...,0.321,0.400,0.721,10,NaN,1,0,0,0,player.fcgi?id=maximi000tha
41,42,Amira Hondras*,18,LAQ,Wmn,NaN,10,38,31,4,...,0.342,0.290,0.632,9,NaN,1,0,1,0,player.fcgi?id=hondra000ami


In [ ]:
# HARDER
# Which team has a higher average OPS; team NYH or SFF?

...

np.float64(0.8987499999999998)